algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+        const int id        = get_global_id(0);
+
+        nodeList[id * nNodeProp + 2] = badVal;
+        nodeList[id * nNodeProp + 3] = badVal;
+    }
+
+    __kernel void doNodesSplit(const __global int * nodeList, int nNodes, __global int * nodeListNew) {
+        const int nNodeProp  = 5; // num of split attributes for node
+        const int badVal     = -1;
+        const int local_id   = get_sub_group_local_id();
+        const int local_size = get_sub_group_size();
+
+        int nCreatedNodes = 0;
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            int splitNode      = (int)(nodeList[i * nNodeProp + 2] != badVal); // featId != -1
+            int newLeftNodePos = nCreatedNodes + sub_group_scan_exclusive_add(splitNode) * 2;
+            if (splitNode)
+            {
+                // split parent node on left and right nodes
+                __global const int * nodeP = nodeList + i * nNodeProp;
+                __global int * nodeL       = nodeListNew + newLeftNodePos * nNodeProp;
+                __global int * nodeR       = nodeListNew + (newLeftNodePos + 1) * nNodeProp;
+
+                nodeL[0] = nodeP[0]; // rows offset
+                nodeL[1] = nodeP[4]; // nRows
+                nodeL[2] = badVal;   // featureId
+                nodeL[3] = badVal;   // featureVal
+                nodeL[4] = nodeP[4]; // num of items in Left part = nRows in new node
+
+                nodeR[0] = nodeL[0] + nodeL[1];
+                nodeR[1] = nodeP[1] - nodeL[1];
+                nodeR[2] = badVal;
+                nodeR[3] = badVal;
+                nodeR[4] = nodeR[1]; // num of items in Left part = nRows in new node
+            }
+            nCreatedNodes += sub_group_reduce_add(splitNode) * 2;
+        }
+    }
+
+    __kernel void splitNodeListOnGroupsBySize(const __global int * nodeList, int nNodes, __global int * bigNodesGroups, __global int * nodeIndeces,
+                                              int minRowsBlock) {
+        /*for now only 3 groups are produced, may be more required*/
+        const int bigNodeLowBorderBlocksNum = 32; // fine, need to experiment and find better one
+        const int blockSize                 = minRowsBlock;
+        const int nNodeProp                 = 5; // num of split attributes for node
+
+        const int local_id   = get_sub_group_local_id();
+        const int nodeId     = get_global_id(1);
+        const int local_size = get_sub_group_size();
+
+        int nBigNodes       = 0;
+        int maxBigBlocksNum = 1;
+        int nMidNodes       = 0;
+        int maxMidBlocksNum = 1;
+
+        /*calculate num of big and mid nodes*/
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            int nRows   = nodeList[i * nNodeProp + 1];
+            int nBlocks = nRows / blockSize + !!(nRows % blockSize);
+            int bigNode = 0;
+            int midNode = 0;
+            if (nBlocks > bigNodeLowBorderBlocksNum)
```

Comment:
It is probably better to initialize this variables using bool to int casts:
`int binNode = (int)(nBlocks > bigNodeLowBorderBlockSum)`
`int midNode = (int)(nBlocks <= binNodeLowBorderBlockSum && nBlocks > 1)`

Reply:
thank you, applied your variant.